### ISSUE TITLE
Rewrite-insert does not invalidate all cache entries correctly

### ISSUE THREAD

**Issue by rene-schoene** [christoff-buerger/racr#77]
In the following situation, RACR does not invalidate the necessary cache entry for the first `B`-node (as this would trigger a re-evaluation of the attribute `uniq-name` on this node.

The grammar (one root with one A and a list of B's):
```scheme
(ast-rule 'Root->A)
(ast-rule 'A->B*<Values)
(ast-rule 'B->timestamp-name)
```

The attributes and two helper functions (`uniq-name` constructs a unique name, whereas `uniq-name2` does the same, but does not cache values for `B`):
```scheme
(define (mk_name n) (cons (ast-node-type n) (ast-child-index n)))
(define (real_parent n)
 (let ((p (ast-parent n)))
  (if (ast-list-node? p) (ast-parent p) p)))

(ag-rule
  uniq-name
  (Root (lambda (n) (list 'Root)))
  (A (lambda (n) (cons (mk_name n) (att-value 'uniq-name (real_parent n)))))
  (B (lambda (n) (cons (mk_name n) (att-value 'uniq-name (real_parent n))))))

(ag-rule
  uniq-name2
  (B #f (lambda (n) (cons (mk_name n) (att-value 'uniq-name (real_parent n))))))
```

AST construction (pseudo-code for a AST with 3 nodes of type B):
```scheme
(define tree
  (create-ast 'Root
    (list (create-ast 'A (list
      (create-ast-list (list
        (create-ast 'B (list 12.0 'first))
        (create-ast 'B (list 13.0 'second))
        (create-ast 'B (list 42.0 'third)))))))))
(rewrite-insert (ast-child 1 (ast-child 1 tree)) 1 (create-ast 'B (list 666 'plus)))
```

Expected behavior:

- all attribute values of `uniq-name` and `uniq-name2` are identical

Actual AST (printed with display-ast and a default printer for the two attributes):
```
-\ Root
 | <uniq-name> {Root}
 |-\ A
 | | <uniq-name> {{A . 1} Root}
 | |-* B
 | | |-\ B
 | | | | <uniq-name2> {{B . 1} {A . 1} Root}
 | | | | <uniq-name> {{B . 1} {A . 1} Root}
 | | | |- timestamp: 666
 | | | |- name: plus
 | | |-\ B
 | | | | <uniq-name2> {{B . 2} {A . 1} Root}
 | | | | <uniq-name> {{B . 1} {A . 1} Root}
 | | | |- timestamp: 12.0
 | | | |- name: first
 | | |-\ B
 | | | | <uniq-name2> {{B . 3} {A . 1} Root}
 | | | | <uniq-name> {{B . 3} {A . 1} Root}
 | | | |- timestamp: 13.0
 | | | |- name: second
 | | |-\ B
 | | | | <uniq-name2> {{B . 4} {A . 1} Root}
 | | | | <uniq-name> {{B . 4} {A . 1} Root}
 | | | |- timestamp: 42.0
 | | | |- name: third
```

Notice the now second node (named `first`) with a bad value for `uniq-name`.

**Comment by christoff-buerger** [comment]
I am on holiday; I will have a look next but one week.

**Comment by christoff-buerger** [comment]
It's a bug because of another intentional bug -- a design decision -- I forgot about, which I think should be changed because ... it's idiotic.

### Short story (that hopefully becomes understandable in the following)

Long time ago I decided, that re-entrance checks are to complicated and re-entrant attributes are unlikely and therefore respective checks should not be performed. I also decided that non-local reference dependencies due to abstract syntax tree nodes as terminal values should be considered (although they are expensive) since bugs due to them are nasty. Then I implemented respective checks for non-local tree-boundary crossings (subtree in- and out-going dependencies). Later I deleted the local checks as they are a subset of the non-local ones, except regarding re-entrance which I already forgot about. Then I forgot about these decisions completely. Then I changed the `ast-child-index` implementation such that it became re-entrant ... for which reason it now is broken :).

### Clarifying example

To understand what a re-entrant attribute is, consider the following cleaned-up example:

```
; This program and the accompanying materials are made available under the
; terms of the MIT license (X11 license) which accompanies this distribution.

; Author: C. Bürger

#!r6rs

(import (rnrs) (racr core) (racr testing))

(define spec (create-specification))

(define (display-ast n)
  (define names
    (let loop ((i (ast-num-children (ast-child 'B* n))))
      (if (> i 0)
          (let ((name  (string-append "name" (number->string i))))
            (cons
             (string->symbol name)
             (cons
              (string->symbol (string-append name "*"))
              (loop (- i 1)))))
          (list))))
  (print-ast
   n
   (map (lambda (a) (cons a (lambda (v) v)))
        (append (list 'index-f 'index-t 'index-t*) names))
   (current-output-port)))

(define (A . b)
  (create-ast spec 'A (list (create-ast-list b))))

(define (B name)
  (create-ast spec 'B (list name)))

(define (index* p c) ; that is how ast-child-index works
  (ast-find-child*
   (lambda (i child)
     (if (eq? child c) i #f))
   p))

(define (name* p i)
  (ast-child 'name (ast-child i p)))

(with-specification
 spec

 (ast-rule 'A->B*)
 (ast-rule 'B->name)
 (compile-ast-specifications 'A)

 ; Re-entrant list-children dependency tests:
 (ag-rule index-f  (B #f (lambda (n)   (index* (ast-parent n) n))))
 (ag-rule index-t  (B    (lambda (n)   (index* (ast-parent n) n))))
 (ag-rule index-t* (B    (lambda (n)   (att-value 'index* (ast-parent n) n))))
 (ag-rule index*   (A    (lambda (n c) (index* (ast-child 'B* n) c))))
 (ag-rule name1    (B    (lambda (n)   (name* (ast-parent n) 1))))
 (ag-rule name2    (B    (lambda (n)   (name* (ast-parent n) 2))))
 (ag-rule name3    (B    (lambda (n)   (name* (ast-parent n) 3))))
 (ag-rule name4    (B    (lambda (n)   (name* (ast-parent n) 4))))
 (ag-rule name5    (B    (lambda (n)   (name* (ast-parent n) 5))))
 (ag-rule name1*   (B    (lambda (n)   (att-value 'name* (ast-parent n) 1))))
 (ag-rule name2*   (B    (lambda (n)   (att-value 'name* (ast-parent n) 2))))
 (ag-rule name3*   (B    (lambda (n)   (att-value 'name* (ast-parent n) 3))))
 (ag-rule name4*   (B    (lambda (n)   (att-value 'name* (ast-parent n) 4))))
 (ag-rule name5*   (B    (lambda (n)   (att-value 'name* (ast-parent n) 5))))
 (ag-rule name*    (A    (lambda (n i) (name* (ast-child 'B* n) i))))
 (compile-ag-specifications))

```

Note, that attribute instances of `index-f`, `index-t` and `name1` to `name5` query nodes up the tree (their parent list nodes) and use them in turn to query down the tree, eventually querying the node they are actually themselves associated with. They do so without using attribute instances up the tree that query down. I call such attributes re-entrant. The starred attributes `index-t*` and `name1*` to `name5*` on the other hand essentially have the same logic, but they use attribute instances up the tree to query down. In consequence they do not directly depend on their node, but only transitively via attribute instances outside of their subtree. They are **not** re-entrant.

Let's play with some list-node rewrites:

```
(let ((ast (A (B 'a) (B 'b) (B 'c) (B 'd))))
  (display-ast ast)
  (rewrite-insert (ast-child 1 ast) 2 (B 'inserted))
  (display-ast ast)
  (rewrite-delete (ast-child 4 (ast-child 1 ast)))
  (display-ast ast))
```

The printouts of the respective attributed abstract syntax trees are:

```
-\ A
 |
 |-* B
 | |
 | |-\ B (B* element)
 | | |@ name4*: d
 | | |@ name3*: c
 | | |@ name2*: b
 | | |@ name1*: a
 | | |@ name4: d
 | | |@ name3: c
 | | |@ name2: b
 | | |@ name1: a
 | | |@ index-t*: 1
 | | |@ index-t: 1
 | | |@ index-f: 1
 | | |> name: a
 | |
 | |-\ B (B* element)
 | | |@ name4*: d
 | | |@ name3*: c
 | | |@ name2*: b
 | | |@ name1*: a
 | | |@ name4: d
 | | |@ name3: c
 | | |@ name2: b
 | | |@ name1: a
 | | |@ index-t*: 2
 | | |@ index-t: 2
 | | |@ index-f: 2
 | | |> name: b
 | |
 | |-\ B (B* element)
 | | |@ name4*: d
 | | |@ name3*: c
 | | |@ name2*: b
 | | |@ name1*: a
 | | |@ name4: d
 | | |@ name3: c
 | | |@ name2: b
 | | |@ name1: a
 | | |@ index-t*: 3
 | | |@ index-t: 3
 | | |@ index-f: 3
 | | |> name: c
 | |
 | |-\ B (B* element)
 | | |@ name4*: d
 | | |@ name3*: c
 | | |@ name2*: b
 | | |@ name1*: a
 | | |@ name4: d
 | | |@ name3: c
 | | |@ name2: b
 | | |@ name1: a
 | | |@ index-t*: 4
 | | |@ index-t: 4
 | | |@ index-f: 4
 | | |> name: d

-\ A
 |
 |-* B
 | |
 | |-\ B (B* element)
 | | |@ name5*: d
 | | |@ name4*: c
 | | |@ name3*: b
 | | |@ name2*: inserted
 | | |@ name1*: a
 | | |@ name5: d
 | | |@ name4: c
 | | |@ name3: b
 | | |@ name2: inserted
 | | |@ name1: a
 | | |@ index-t*: 1
 | | |@ index-t: 1
 | | |@ index-f: 1
 | | |> name: a
 | |
 | |-\ B (B* element)
 | | |@ name5*: d
 | | |@ name4*: c
 | | |@ name3*: b
 | | |@ name2*: inserted
 | | |@ name1*: a
 | | |@ name5: d
 | | |@ name4: c
 | | |@ name3: b
 | | |@ name2: inserted
 | | |@ name1: a
 | | |@ index-t*: 2
 | | |@ index-t: 2
 | | |@ index-f: 2
 | | |> name: inserted
 | |
 | |-\ B (B* element)
 | | |@ name5*: d
 | | |@ name4*: c
 | | |@ name3*: b
 | | |@ name2*: inserted
 | | |@ name1*: a
 | | |@ name5: d
 | | |@ name4: c
 | | |@ name3: b
 | | |@ name2: b
 | | |@ name1: a
 | | |@ index-t*: 3
 | | |@ index-t: 2
 | | |@ index-f: 3
 | | |> name: b
 | |
 | |-\ B (B* element)
 | | |@ name5*: d
 | | |@ name4*: c
 | | |@ name3*: b
 | | |@ name2*: inserted
 | | |@ name1*: a
 | | |@ name5: d
 | | |@ name4: c
 | | |@ name3: c
 | | |@ name2: inserted
 | | |@ name1: a
 | | |@ index-t*: 4
 | | |@ index-t: 4
 | | |@ index-f: 4
 | | |> name: c
 | |
 | |-\ B (B* element)
 | | |@ name5*: d
 | | |@ name4*: c
 | | |@ name3*: b
 | | |@ name2*: inserted
 | | |@ name1*: a
 | | |@ name5: d
 | | |@ name4: d
 | | |@ name3: b
 | | |@ name2: inserted
 | | |@ name1: a
 | | |@ index-t*: 5
 | | |@ index-t: 5
 | | |@ index-f: 5
 | | |> name: d

-\ A
 |
 |-* B
 | |
 | |-\ B (B* element)
 | | |@ name4*: d
 | | |@ name3*: b
 | | |@ name2*: inserted
 | | |@ name1*: a
 | | |@ name4: d
 | | |@ name3: b
 | | |@ name2: inserted
 | | |@ name1: a
 | | |@ index-t*: 1
 | | |@ index-t: 1
 | | |@ index-f: 1
 | | |> name: a
 | |
 | |-\ B (B* element)
 | | |@ name4*: d
 | | |@ name3*: b
 | | |@ name2*: inserted
 | | |@ name1*: a
 | | |@ name4: d
 | | |@ name3: b
 | | |@ name2: inserted
 | | |@ name1: a
 | | |@ index-t*: 2
 | | |@ index-t: 2
 | | |@ index-f: 2
 | | |> name: inserted
 | |
 | |-\ B (B* element)
 | | |@ name4*: d
 | | |@ name3*: b
 | | |@ name2*: inserted
 | | |@ name1*: a
 | | |@ name4: d
 | | |@ name3: b
 | | |@ name2: b
 | | |@ name1: a
 | | |@ index-t*: 3
 | | |@ index-t: 2
 | | |@ index-f: 3
 | | |> name: b
 | |
 | |-\ B (B* element)
 | | |@ name4*: d
 | | |@ name3*: b
 | | |@ name2*: inserted
 | | |@ name1*: a
 | | |@ name4: d
 | | |@ name3: b
 | | |@ name2: inserted
 | | |@ name1: a
 | | |@ index-t*: 4
 | | |@ index-t: 4
 | | |@ index-f: 4
 | | |> name: d
```

As can be seen, the starred attributes `index-t*` and `name1*` to `name5*` are working whereas their re-entrant counterparts are broken. Note, that the re-entrant attributes are only broken for the node-position actually touched by a rewrite; they are only broken for the list-child that was at the position a `rewrite-inserted` inserted a new node into (likewise for `rewrite-delete`). They work otherwise, thanks to tree-boundary crossing checks (for example the `index-t` attribute of the 3rd child depends on the 2nd child, which is detected when a new 2nd child is inserted (tree-boundary crossing dependency from the `index-t` instance of the 3rd child to the 2nd), resulting in a respective cache flush.

### Tree-boundary crossing checks

All our list-node rewriting problems boil down to check if attribute dependencies cross the tree-boundaries of list-elements. When we know, if some attribute instance depends on some information within the subtree spanned by a list-element, we know if it must be flushed due to some list change (element insertion, addition or deletion).

Such a tree-boundary crossing check is local if it just considers the attributes of the list-element (the node spanning the tree checked for crossings) and its respective list-node (its parent). Otherwise it is non-local, in which case crossing checks require the subtree to be traversed and the origins of all dependencies on all its attributes have to be be checked if within the subtree or not. This requires for each origin a tree traversal upwards the chain of parents of the respective boundary. Thus, non-local tree-boundary crossing checks are expensive.

I introduced non-local tree-boundary crossing checks nevertheless to handle non-local reference dependencies due to abstract syntax tree nodes provided by terminals. To understand the issue consider how an attribute instance outside of the subtree of some node can actually depend on information within it (and vice versa, how some attribute instance within a subtree can depend on information outside of it).

There are two cases:
 * (1) The attribute instance queries the node with the respective information directly via `ast-child`/`get-parent` using the node itself is associated with.
 * (2) The attribute instance somehow got a node reference to query.

Case (1) obviously is local and therefore trivial to check runtime-performance wise. Case (2) can be due to two possibilities:
 * (2.1) A reference attribute was used.
 * (2.2) The node was given, either:
   * (2.2.1) as an attribute argument or
   * (2.2.2) by a queried terminal (i.e., the queried terminal's value was a remote non-terminal node of the abstract syntax tree).

In cases (2.1) and (2.2.1) dependencies are properly handled by the queried reference attribute and the providing calling attribute respectively. Thus, only cases (1) and (2.2.2) have to be checked for respective cache flushing, whereas case (1) is local. Only case (2.2.2) requires non-local tree-boundary crossing checks.

### Redesign

At the moment I do not check if an attribute instance locally depends on its own node via a query down-wards the tree (i.e., via `ast-child`). This is a re-entrant check and trivial to do, also performance-wise. I already track the dependency on the node (i.e., the attribute instance knows that it explicitly queried its own node); the only pice of information missing for proper re-entrant checks is if it queried its node down- or upwards (i.e., via `ast-child` or `get-parent`). This information can be easily incorporated to the dependencies added by `ast-child` and `ast-parent`.

I also think, that I should drop case (2.2.2) and instead only stick with local tree-boundary crossing checks. The reason to drop non-local tree-boundary crossing checks is that _RACR_ does not -- and cannot in general -- investigate terminal values. Side-effects that change the state of a _Scheme_ entity used as a terminal value cannot be dependency-tracked; it is the responsibility of developers to rewrite the respective terminal whenever the state of its stateful value changes. To request respective terminal rewrites in case abstract syntax tree nodes are used as terminal values is just in line with this policy.

*I will reread my blob of text tomorrow and fix it required*

**Comment by christoff-buerger** [comment]
Commit 13fc68dc6982055fd96217fbd24250937f53f169, which fixed issue #78, also fixes this issue.
